pkg/apis/v1beta1/nodepool_budgets_test.go

```diff
@@ -70,30 +95,56 @@ var _ = Describe("Budgets", func() {
+			budgets[5].Schedule = lo.ToPtr("* * * * *")
+			budgets[5].Duration = lo.ToPtr(metav1.Duration{Duration: lo.Must(time.ParseDuration("1h"))})
+
+			disruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, fakeClock, 100)
+			Expect(err).To(BeNil())
+			Expect(disruptionsByReason[DisruptionReasonUnderutilized]).To(Equal(0))
+			Expect(disruptionsByReason[DisruptionReasonDrifted]).To(Equal(0))
+			Expect(disruptionsByReason[DisruptionReasonEmpty]).To(Equal(0))
+			Expect(disruptionsByReason[DisruptionReasonExpired]).To(Equal(0))
+		})
+		It("should return MaxInt32 for all reasons when there are no active budgets", func() {
+			for i := range budgets {
+				budgets[i].Schedule = lo.ToPtr("@yearly")
+				budgets[i].Duration = lo.ToPtr(metav1.Duration{Duration: lo.Must(time.ParseDuration("1h"))})
+			}
+			disruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, fakeClock, 100)
+			Expect(err).To(BeNil())
+
+			// All budgets should have unbounded disruptions when inactive
+			for _, disruptions := range disruptionsByReason {
+				Expect(disruptions).To(Equal(math.MaxInt32))
+			}
 		})
-		It("should return MaxInt32 if all crons are inactive", func() {
-			budgets[0].Schedule = lo.ToPtr("@yearly")
-			budgets[1].Schedule = lo.ToPtr("@yearly")
-			budgets[2].Schedule = lo.ToPtr("@yearly")
+		It("should ignore reason-defined budgets when inactive", func() {
 			budgets[3].Schedule = lo.ToPtr("@yearly")
-			min := nodePool.MustGetAllowedDisruptions(ctx, fakeClock, 100)
-			Expect(min).To(BeNumerically("==", math.MaxInt32))
-		})
-		It("should return zero values if a schedule is invalid", func() {
-			budgets[0].Schedule = lo.ToPtr("@wrongly")
-			min := nodePool.MustGetAllowedDisruptions(ctx, fakeClock, 100)
-			Expect(min).To(BeNumerically("==", 0))
+			budgets[4].Schedule = lo.ToPtr("@yearly")
+			disruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, fakeClock, 100)
+			Expect(err).To(BeNil())
+			for _, disruptions := range disruptionsByReason {
+				Expect(disruptions).To(Equal(10))
+			}
+		})
+		It("should return the budget for all disruption reasons when undefined", func() {
+			nodePool.Spec.Disruption.Budgets = budgets[:1]
+			Expect(len(nodePool.Spec.Disruption.Budgets)).To(Equal(1))
+			disruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, fakeClock, 100)
+			Expect(err).To(BeNil())
+			Expect(len(budgets[0].Reasons)).To(Equal(0))
+			for _, disruptions := range disruptionsByReason {
+				Expect(disruptions).To(Equal(10))
+			}
+		})
+		It("should get the minimum budget for each reason", func() {
+			disruptionsByReason, err := nodePool.GetAllowedDisruptionsByReason(ctx, fakeClock, 100)
+			Expect(err).To(BeNil())
+			Expect(disruptionsByReason[DisruptionReasonDrifted]).To(Equal(5))
+			// The budget where reason == nil overrides the budget with a specified reason
+			Expect(disruptionsByReason[DisruptionReasonUnderutilized]).To(Equal(10))
 		})
```

Comment:
Can you add more to this test that just makes sure that each disruption reason, if more restrictive than the non-reason allowed budget, applies as well just for full coverage? 

This would be adding in the same budget with empty and expired.

Reply:
added that coverage good call.